### PR TITLE
[refactor] [CUDA] Wrap the default profiling tool as EventToolkit , add a new class for CUPTI toolkit

### PR DESCRIPTION
**Pull request by yolo2themoon**
Related issue = #2909 #2902

The implementation of CuptiToolkit will be added in the next PR.

The code in sync() is now more abstract:
```
void KernelProfilerCUDA::sync() {
  // sync
  CUDADriver::get_instance().stream_synchronize(nullptr);

  // update
  if (tool_ == ProfilingToolkit::event) {
    event_toolkit_->update_record(traced_records_);
    event_toolkit_->update_timeline(traced_records_);
    statistics_on_traced_records();
    event_toolkit_->clear();
  } else if (tool_ == ProfilingToolkit::cupti) {
    cupti_toolkit_->calculate_metric_values();
    statistics_on_traced_records();
    cupti_toolkit_->end_profiling();
    cupti_toolkit_->deinit_cupti();  
    // reinit is necessary to get accurate  result
    cupti_toolkit_->init_cupti();  // data image will be cleared
    cupti_toolkit_->begin_profiling();
  }
}
```

### CODE REVIEW

File: taichi/backends/cuda/cuda_profiler.h [hunk 1]
```diff
@@ -1,40 +1,72 @@
...
   void stop(KernelProfilerBase::TaskHandle handle) override;
 
+  bool statistics_on_traced_records();
+
+  KernelProfilerBase::TaskHandle start_with_handle(
+      const std::string &kernel_name) override;
+
  private:
-  void *base_event_{nullptr};
+  ProfilingToolkit tool_ = ProfilingToolkit::event;
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.h:42]: nit: will be nice to add a note to highlight an invariant here:
tool_ == event -> event_toolkit_ not nullptr
tool_ == event-> cupti_toolkit_ not nullptr.

File: taichi/backends/cuda/cuda_profiler.cpp [hunk 2]
```diff
@@ -1,19 +1,167 @@
...
+
+KernelProfilerCUDA::KernelProfilerCUDA() {
+// if Taichi was compiled with CUDA toolit, then use CUPTI
+// TODO : add set_mode() to select toolkit by user
+#if defined(TI_WITH_CUDA_TOOLKIT)
+  if (check_device_capability())
+    tool_ = ProfilingToolkit::cupti;
+#endif
+
+  if (tool_ == ProfilingToolkit::event) {
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.cpp:32]: nit: Looks like this if can be merged with the `#if` above?

File: taichi/backends/cuda/cuda_profiler.cpp [hunk 3]
```diff
@@ -1,19 +1,167 @@
...
+  if (tool_ == ProfilingToolkit::event)
+    task_handle = event_toolkit_->start_with_handle(task_name);
+  else if (tool_ == ProfilingToolkit::cupti) {
+    TI_NOT_IMPLEMENTED;
+  }
+}
+
+void KernelProfilerCUDA::stop(KernelProfilerBase::TaskHandle handle) {
+  if (tool_ == ProfilingToolkit::event)
+    CUDADriver::get_instance().event_record(handle, 0);
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.cpp:56]: what happens to cupti in this case?

File: taichi/backends/cuda/cuda_profiler.cpp [hunk 4]
```diff
@@ -1,19 +1,167 @@
...
+  return true;
+}
+
+void KernelProfilerCUDA::sync() {
+  // sync
+  CUDADriver::get_instance().stream_synchronize(nullptr);
+
+  // update
+  if (tool_ == ProfilingToolkit::event) {
+    event_toolkit_->update_record(traced_records_);
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.cpp:83]: wrap these into a EventToolkit::sync()?

File: taichi/backends/cuda/cuda_profiler.cpp [hunk 5]
```diff
@@ -1,19 +1,167 @@
...
+    event_toolkit_->update_timeline(traced_records_);
+    statistics_on_traced_records();
+    event_toolkit_->clear();
+  } else if (tool_ == ProfilingToolkit::cupti) {
+    cupti_toolkit_->calculate_metric_values();
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.cpp:88]: ditto

**Review by ailzhang — APPROVED** [review]
Looks great! Thanks!

File: taichi/backends/cuda/cuda_profiler.cpp (hunk 3 above)
Comment (yolo2themoon) [reply taichi/backends/cuda/cuda_profiler.cpp:56]: `cuEvent` is controlled by host(CPU), needs to call star() and stop() explicitly.
With `CUPTI` , profiling is performed by device(GPU), just need to evaluate metric value form the CUPTI data image while `sync()`

File: taichi/backends/cuda/cuda_profiler.cpp [hunk 6]
```diff
@@ -1,19 +1,167 @@
...
+        "7.0 , fallback to default kernel profiler");
+    return false;
+  }
+  return true;
+}
+
+KernelProfilerCUDA::KernelProfilerCUDA() {
+// if Taichi was compiled with CUDA toolit, then use CUPTI
+// TODO : add set_mode() to select toolkit by user
+#if defined(TI_WITH_CUDA_TOOLKIT)
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_profiler.cpp:27]: Btw with this logic here, is cudaEvent still the default profiling method (as described in the comment)? Looks like if I install cuda toolkit, my default is cupti?

File: taichi/backends/cuda/cuda_profiler.cpp (hunk 4 above)
Comment (yolo2themoon) [reply taichi/backends/cuda/cuda_profiler.cpp:83]: thanks! I'll get this done in the next PR today

File: taichi/backends/cuda/cuda_profiler.cpp (hunk 6 above)
Comment (yolo2themoon) [reply taichi/backends/cuda/cuda_profiler.cpp:27]: one more prerequisite in needed ( compile Taichi with ):
`TAICHI_CMAKE_ARGS="-DTI_WITH_CUDA_TOOLKIT:BOOL=ON" python3 setup.py develop --user`

Comment (yolo2themoon) [reply taichi/backends/cuda/cuda_profiler.cpp:27]: some fallback logic will be added in the next PR : )

Comment (yolo2themoon) [reply taichi/backends/cuda/cuda_profiler.cpp:27]: `TI_WITH_CUDA_TOOLKIT` now means `TI_COMPILE_WITH_CUDA_TOOLKIT`
I'll refactor the logic here after migrate to `FindCUDA.cmake`  #2899 ( not done yet )

Comment (ailzhang) [reply taichi/backends/cuda/cuda_profiler.cpp:29]: SG!

**Pull request merged**